source/extensions/filters/http/composite/filter.cc

```diff
@@ -0,0 +1,192 @@
+                              Http::FilterHeadersStatus::Continue, headers, end_stream);
+}
+
+Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_stream) {
+  return delegateFilterAction(delegated_filter_, &StreamDecoderFilter::decodeData,
+                              Http::FilterDataStatus::Continue, data, end_stream);
+}
+
+Http::FilterTrailersStatus Filter::decodeTrailers(Http::RequestTrailerMap& trailers) {
+  return delegateFilterAction(delegated_filter_, &StreamDecoderFilter::decodeTrailers,
+                              Http::FilterTrailersStatus::Continue, trailers);
+}
+
+Http::FilterMetadataStatus Filter::decodeMetadata(Http::MetadataMap& metadata_map) {
+  return delegateFilterAction(delegated_filter_, &StreamDecoderFilter::decodeMetadata,
+                              Http::FilterMetadataStatus::Continue, metadata_map);
+}
+
+void Filter::decodeComplete() {
+  if (delegated_filter_) {
+    delegated_filter_->decodeComplete();
+  }
+}
+
+Http::FilterHeadersStatus Filter::encode100ContinueHeaders(Http::ResponseHeaderMap& headers) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encode100ContinueHeaders,
+                              Http::FilterHeadersStatus::Continue, headers);
+}
+
+Http::FilterHeadersStatus Filter::encodeHeaders(Http::ResponseHeaderMap& headers, bool end_stream) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encodeHeaders,
+                              Http::FilterHeadersStatus::Continue, headers, end_stream);
+}
+Http::FilterDataStatus Filter::encodeData(Buffer::Instance& data, bool end_stream) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encodeData,
+                              Http::FilterDataStatus::Continue, data, end_stream);
+}
+Http::FilterTrailersStatus Filter::encodeTrailers(Http::ResponseTrailerMap& trailers) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encodeTrailers,
+                              Http::FilterTrailersStatus::Continue, trailers);
+}
+Http::FilterMetadataStatus Filter::encodeMetadata(Http::MetadataMap& metadata_map) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encodeMetadata,
+                              Http::FilterMetadataStatus::Continue, metadata_map);
+}
+void Filter::encodeComplete() {
+  if (delegated_filter_) {
+    delegated_filter_->encodeComplete();
+  }
+}
+void Filter::onMatchCallback(const Matcher::Action& action) {
+  const auto& composite_action = action.getTyped<ExecuteFilterAction>();
+
+  FactoryCallbacksWrapper wrapper(*this);
+  composite_action.createFilters(wrapper);
+
+  if (!wrapper.errors_.empty()) {
+    stats_.filter_delegation_error_.inc();
+    ENVOY_LOG(debug, "failed to create delegated filter {}",
+              accumulateToString<absl::Status>(
+                  wrapper.errors_, [](const auto& status) { return status.ToString(); }));
+    return;
+  }
```

Comment:
Maybe this is an issue with the matching engine that needs to be fixed, but shouldn't this return an error code which should tell the engine to actually do something like fail the request? We can discuss later about fail open vs. fail closed, but it seems like it shouldn't just be a nop? wdyt?

Reply:
Yeah that's a good point, I think in the short term we could local reply here with a 500 or something, but maybe it would be nicer with direct integration with the matching framework? Or maybe just a success bool so that we can avoid calling decodeHeaders etc. if matching resulted in a local reply?

Reply:
IMO I would have it be built-in to the matching engine with a return code vs. void, but fine if you want to TODO here to fix that later and for now just do a local reply.

Reply:
Turns out we can't do a local reply here since the associated filter callback is invoked: 
```
[2021-04-12 20:53:43.183][48][error][envoy_bug] [source/common/http/filter_manager.cc:523] envoy bug failure: !continue_iteration || !state_.local_complete_. Details: Filter did not return StopAll or StopIteration after sending a local reply.
```

Seems like we need some HCM changes to make this work, wdyt about landing this as is with a TODO and then follow up once we improve the onMatchCallback API?

Reply:
Sure that's fine, thanks for checking.